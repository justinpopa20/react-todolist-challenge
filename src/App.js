import React, { Component } from "react";
import { connect } from "react-redux"
import { addItem } from "./actions"

import Input from "./components/input";
import Button from "./components/button";

import List from "./containers/TodoList";

import "./App.css";

export class App extends Component {
    state = { todoText: "" }

    handleInputChange = todoText => {
        this.setState({ todoText })
    }

    handleAddTodo = () => {
        const { todoText } = this.state
        if (!todoText) return

        this.props.addItem(todoText)
        this.setState({ todoText: "" })
    }

    render() {
        return (
            <div className="container">

                <h1>Todo List</h1>

                <div className="add-item-to-list">
                    <Input
                        name="item"
                        data-testid="add-input"
                        placeholder="New Item..."
                        value={this.state.todoText}
                        onChange={this.handleInputChange}
                    />
                    <Button onClick={this.handleAddTodo} type="add" data-testid="add-btn">
                        Add
                    </Button>
                </div>
                <div className="error-container">{this.props.error}</div>
                <List />
            </div>
        );
    }
}

const mapStateToProps = state => ({
    error: state.list.error
})

const mapDispatchToProps = dispatch => ({
    addItem: text => dispatch(addItem(text))
})

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(App)
