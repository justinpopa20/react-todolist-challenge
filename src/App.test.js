import * as React from 'react'
import { render, fireEvent } from '@testing-library/react'
import rootReducer from './reducers'
import thunk from 'redux-thunk';
import { createStore, applyMiddleware } from 'redux'
import { Provider } from 'react-redux'

import { default as ConnectedApp } from './App'
import { App } from './App'

function renderWithRedux(
  ui,
  { initialState, store = createStore(rootReducer, initialState, applyMiddleware(thunk)) } = {}
) {
  return {
    ...render(<Provider store={store}>{ui}</Provider>),
    store,
  }
}

describe('App', () => {
  const props = {
    addItem: jest.fn(),
  }

  beforeEach(() => {
    jest.clearAllMocks()
  })

  test('should render - smoke test', () => {
    const wrapper = renderWithRedux(<ConnectedApp />)

    expect(wrapper.baseElement).toMatchSnapshot()
  })

  test('add todo', () => {
    const wrapper = renderWithRedux(<App {...{...props}} />)
    const input = wrapper.getByTestId("add-input")
    const addBtn = wrapper.getByTestId("add-btn")

    fireEvent.change(input, { target: { value: 'test123' } })
    fireEvent.click(addBtn)

    expect(props.addItem).toHaveBeenCalledWith('test123')
  })

  test('not add empty todo', () => {
    const wrapper = renderWithRedux(<App {...{...props}} />)
    const addBtn = wrapper.getByTestId("add-btn")

    fireEvent.click(addBtn)

    expect(props.addItem).not.toHaveBeenCalled()
  })

  test('display errors', () => {
    const errorText = 'Problem loading list'
    const wrapper = renderWithRedux(<App {...{...props, error: errorText}} />)
    const errorEl = wrapper.getByText(errorText)

    expect(errorEl.textContent).toBe(errorText)
  })
})