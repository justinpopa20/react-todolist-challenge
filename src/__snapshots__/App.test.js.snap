// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should render - smoke test 1`] = `
<body>
  <div>
    <div
      class="container"
    >
      <h1>
        Todo List
      </h1>
      <div
        class="add-item-to-list"
      >
        <input
          type="text"
          name="item"
          placeholder="New Item..."
          data-testid="add-input"
          value=""
        />
        <button
          class="button add"
          data-testid="add-btn"
        >
          Add
        </button>
      </div>
      <div
        class="error-container"
      />
      <table>
        <thead>
          <tr>
            <th
              width="66%"
            >
              Item
            </th>
            <th>
              Actions
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td>
              Loading...
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</body>
`;
